api/session_token_login_provider.go

```diff
@@ -73,11 +75,19 @@ func (p *sessionTokenLoginProvider) Login(ctx context.Context, caller base.APICa
 	return nil, errors.Trace(err)
 }
 
-func (p *sessionTokenLoginProvider) initiateDeviceLogin(ctx context.Context, caller base.APICaller) error {
-	if p.printOutputFunc == nil {
+func (p *sessionTokenLoginProvider) printOutput(format string, params ...any) error {
+	if p.cmdOut == nil {
 		return errors.New("cannot present login details")
 	}
+	message := fmt.Sprintf(format, params...)
+	if len(message) > 0 && message[len(message)-1] != '\n' {
```

Comment:
This seems error prone.

Reply:
I actually took it from Juju's logger https://github.com/juju/loggo/blob/v1.0.0/logger.go#L137
I would ask if you can think of a situation where it might be problematic?
Also switching to `strings.HasSuffix(message, '"\n")` could be clearer.